src/Estimators/NESpaceGrid.cpp

```diff
@@ -588,16 +588,31 @@ void NESpaceGrid<REAL>::accumulate(const ParticlePos& R,
       {
         particles_outside[p] = false;
         Point u              = dot(axinv_, (R[p] - origin_));
+	// numerical accuracy of the dot is not good enough for this to be not produce out of bounds indexes
+	// with REAL = float, u 10^-8 outside of umin_ to  umax_ is not uncommon i.e  p > (1 in 32000)
+	// for the coordinate ranges typical of our simulations. I guess the branch predictor will mostly
+	// mitigate this.
+	auto gmapIndex = [this](int d, const auto& u) { int raw_index = floor((u[d] - this->umin_[d]) * this->odu_[d]);
```

Comment:
Is `u[d] - this->umin_[d] ` always positive? I'm wondering if raw_index can be `-5` when the chosen space is located in the center of a lattice.

Reply:
This code is just a direct port from legacy and it is complicated in how it sets up the grid lattice coordinate system, I think it might take more time than I can give today to trace through it all. So here is the current case:
```xml
    <parameter name='lattice' units='bohr'>
        3.37316115        3.37316115        0.00000000
        0.00000000        3.37316115        3.37316115
        3.37316115        0.00000000        3.37316115
    </parameter>
```
applyMinimumImage guarantees I will get something that is positive definite for my r_particle

With the space grid input like
```xml
           <spacegrid coord="cartesian">
             <origin p1="zero"/>
             <axis p1="a1" scale=".5" label="x" grid="-1 (.1) 1"/>
             <axis p1="a2" scale=".5" label="y" grid="-1 (.1) 1"/>
             <axis p1="a3" scale=".5" label="z" grid="-1 (.1) 1"/>
           </spacegrid>
``` 
This gives us a transformation matrix that looks like
```c++
axinv_ = {0.29645781969236779, 0.29645781969236779, -0.29645781969236779, -0.29645781969236779, 0.29645781969236779, 0.29645781969236779,
    0.29645781969236779, -0.29645781969236779, 0.29645781969236779}
```
Which seems to be based on the hard assumption that if `coord="cartesian"` whatever grid is defined it going to map the unit cell into it.

Which ideally would take all r_particle to u_grid_particle = {[-1,1],[-1,1],[-1,1]} but instead takes us to something that has some small numerical error +/- for floatingpoint operations of fixed precision.

(I find the way the origin argument works here counterintuitive because I think we are really getting a grid whose origin is `{0.5,0.5,0.5}` in the actual lattice coordinates.  But the Energy Density Estimator grid docs do say: "zero -- Cell center" so I guess this is a feature, not what I would do here since both seems a totally valid way to define the lattice.)

then we get to the actual gmapIndex algorithm. `u - umin` where `umin = {-1,-1,-1}` Which gives a a positive definite [0,2) * odu_

The legacy code has very long and complex initialization that appears to guarantee that the relation between umin_, umax_ and odu_ is always such  without floating point error the indexes are always [0, umax[d]-umin[d] * (1 / step[d]) ).  There was no unit testing of this and I certainly have not had the time to test corner cases I doubt it is terribly robust, but perhaps I just don't see that it is perfection outside of the naivete wrt float point error.

All cartesian coordinate are forced to be treated as period 
```
bool isPeriodic() const { return !(coord_form_ == CoordForm::CYLINDRICAL || coord_form_ == CoordForm::SPHERICAL); }
```

A concern here is  minimumImage is only called in the EDE if the lattice is not open so that's a corner case that is not handled.

cartesian grid coordinates must also be symmetric about whatever origin is used.  None of this is checked for more extensively in legacy than in the current code.  In the scope of this PR I don't think its feasible to do more than fix the out of bounds in the cases that can occur when the user sets up their grid correctly.  Otherwise the exception gives them what we can to fix the grid.